config_ext_test.go

```diff
@@ -0,0 +1,350 @@
+			Value:    val,
+			HasValue: true,
+			Interface: map[interface{}]interface{}{
+				"honda":  "civic",
+				"toyota": "camry",
+				"nissan": "altima",
+				"ford":   "fiesta",
+			},
+			Populate: &m,
+		})
+		assert.Equal(t, map[string]string{
+			"honda":  "civic",
+			"toyota": "camry",
+			"nissan": "altima",
+			"ford":   "fiesta",
+		}, m, "unexpected result after populate")
+	})
+
+	t.Run("default replaces sequences", func(t *testing.T) {
+		var s []string
+		val, err := provider.Get("fun").WithDefault([]string{"delorean"})
+		require.NoError(t, err, "couldn't set default")
+		run(TestCase{
+			Value:     val,
+			HasValue:  true,
+			Interface: []interface{}{"maserati", "lamborghini"},
+			Populate:  &s,
+		})
+		assert.Equal(t, []string{"maserati", "lamborghini"}, s, "unexpected result after populate")
+	})
+
+	t.Run("default ignores type mismatches", func(t *testing.T) {
+		// TODO: We should return an error in strict mode.
+		// fun is a sequence, so we shouldn't be able to merge in a mapping
+		val, err := provider.Get("fun").WithDefault(map[string]string{"dodge": "viper"})
+		assert.NoError(t, err, "error on type mismatch in permissive mode")
+		assert.Equal(t, []interface{}{"maserati", "lamborghini"}, val.Value(), "wrong merged output")
+	})
+
+	t.Run("chained defaults", func(t *testing.T) {
+		// Defaults should be FIFO: once we've set a default value, we can't later
+		// change our minds and set a different default.
+		val, err := provider.Get("top").WithDefault(map[string]string{"middle": "bottom"})
+		require.NoError(t, err, "couldn't set first default")
+		val, err = val.Get("middle").WithDefault("boom")
+		require.NoError(t, err, "couldn't set second default")
+
+		var s string
+		run(TestCase{
+			Value:     val,
+			HasValue:  true,
+			Interface: "bottom",
+			Populate:  &s,
+		})
+		assert.Equal(t, "bottom", s, "unexpected result after populate")
+	})
+
+	t.Run("deep copy", func(t *testing.T) {
+		t.Skip("TODO: mutations are visible to other callers")
+
+		const key = "foobar"
+		unmarshal := func() map[interface{}]interface{} {
+			var m map[string]interface{}
+			require.NoError(t, provider.Get(Root).Populate(&m), "Populate failed")
+			return m["practical"].(map[interface{}]interface{})
+		}
+
+		before := unmarshal()
+		require.NotContains(t, before, key, "precondition failed: key %q already in map", key)
+		before[key] = "bazbing"
+
+		after := unmarshal()
+		assert.NotContains(t, after, key, "didn't deep copy config")
+	})
+
+	t.Run("named field", func(t *testing.T) {
+		type cfg struct {
+			Boring []string `yaml:"fun"`
+		}
+
+		var c cfg
+		require.NoError(t, provider.Get(Root).Populate(&c))
+		assert.Equal(t, 2, len(c.Boring), "wrong number of fun cars")
+	})
+
+	t.Run("ignored field", func(t *testing.T) {
+		type cfg struct {
+			Boring []string `yaml:"-"`
+		}
+
+		var c cfg
+		require.NoError(t, provider.Get(Root).Populate(&c))
+		assert.Equal(t, 0, len(c.Boring), "wrong number of fun cars")
```

Comment:
optionally: `assert.Empty`

Reply:
For no particular reason, I usually avoid that helper. Alex mentioned the same thing, so I'll change.